app/lib/encryption/encryption.js

```diff
@@ -0,0 +1,420 @@
+import EJSON from 'ejson';
+import SimpleCrypto from 'react-native-simple-crypto';
+import { Q } from '@nozbe/watermelondb';
+
+import {
+	toString,
+	utf8ToBuffer,
+	splitVectorData,
+	joinVectorData,
+	randomPassword
+} from './utils';
+import {
+	E2E_PUBLIC_KEY,
+	E2E_PRIVATE_KEY,
+	E2E_RANDOM_PASSWORD_KEY,
+	E2E_STATUS,
+	E2E_MESSAGE_TYPE,
+	E2E_ROOM_TYPES
+} from './constants';
+import RocketChat from '../rocketchat';
+import EncryptionRoom from './encryption.room';
+import UserPreferences from '../userPreferences';
+import database from '../database';
+import protectedFunction from '../methods/helpers/protectedFunction';
+
+class Encryption {
+	constructor() {
+		this.ready = false;
+		this.privateKey = null;
+		this.roomInstances = {};
+	}
+
+	// Initialize Encryption client
+	initialize = () => {
+		this.roomInstances = {};
+
+		// Don't await these promises
+		// so these can run parallelized
+		this.decryptPendingSubscriptions();
+		this.decryptPendingMessages();
+
+		// Mark Encryption client as ready
+		this.ready = true;
+	}
+
+	// Stop Encryption client
+	stop = () => {
```

Comment:
now that we have a "ready state", we can cancel ongoing decryptions/encryptions that are waiting to the client be ready, if it was ready at the moment of the decrypt, so this will be finished successfully and we don't need to cancel.